### PR TITLE
Boundary bugfix(es) and exposing `FluxDiv_` interface

### PR DESCRIPTION
**Pull request by pgrete**
## PR Summary

Also adjust `dt` calc as the original version resulted in a `nan` when it was initially set to `max`

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [ ] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

File: src/driver/driver.cpp
Comment (Yurlungur) [inline src/driver/driver.cpp:161]: Wow... you hit this? I'm impressed.

File: src/bvals/boundary_conditions.cpp
Comment (Yurlungur) [inline src/bvals/boundary_conditions.cpp]: Good catch!

File: src/driver/driver.cpp
Comment (pgrete) [reply src/driver/driver.cpp:161]: Tracking down another problem I used 
```
#include <fenv.h> 
feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW);
```
and immediately found this one directly during init

Comment (Yurlungur) [reply src/driver/driver.cpp:161]: heh nice

File: src/interface/update.hpp
Comment (Yurlungur) [inline src/interface/update.hpp]: We should consider renaming `FluxDiv_` to `FluxDiv` or `FluxDivHelper` or something if it's publicly exposed.

**Comment by Yurlungur** [comment]
@pgrete Changing k-bounds initialized the uninitialized variables in the ghost zones at the initial time. I just pushed to your branch.

File: src/interface/update.hpp
Comment (pgrete) [reply src/interface/update.hpp]: Done.

**Review by Yurlungur — APPROVED** [review]

**Comment by pgrete** [comment]
What do you think about updating the advection example so that the background has a value of "1" versus "0" currently and then I'd add a `PARTHENON_REQUIRE(advected != 0)` to the `SquareIt` function that is enrolled as `FillDerived`?
This would mean updating the regression data, but I think it's worth adding as is causes almost no overhead an will increase test coverage (namely, that all cells are always initialized).

**Comment by Yurlungur** [comment]
> What do you think about updating the advection example so that the background has a value of "1" versus "0" currently and then I'd add a `PARTHENON_REQUIRE(advected != 0)` to the `SquareIt` function that is enrolled as `FillDerived`?
> This would mean updating the regression data, but I think it's worth adding as is causes almost no overhead an will increase test coverage (namely, that all cells are always initialized).

I think this is a good idea, so long as the advected field now has whatever form it currently has + 1. ;)

**Review by jlippuner — APPROVED** [review]
Looks great and great catch!

Since we're making changes to the boundary conditions file anyway, I thought this would be a good opportunity to reduce some code duplication. I've made one generic boundary condition function covering all 12 cases (X1, X2, X3) x (inner, outer) x (outflow, reflection). Hopefully, this would make it easier to avoid such copy&paste errors in the future (if the boundary condition code needs to change). Please let me know what you think.

**Comment by Yurlungur** [comment]
> Looks great and great catch!
> 
> Since we're making changes to the boundary conditions file anyway, I thought this would be a good opportunity to reduce some code duplication. I've made one generic boundary condition function covering all 12 cases (X1, X2, X3) x (inner, outer) x (outflow, reflection). Hopefully, this would make it easier to avoid such copy&paste errors in the future (if the boundary condition code needs to change). Please let me know what you think.

In principal, I'm in favor of reduced code, but I worry that I can't quite parse whether or not the generic version covers all 12 specialized versions. I'm not sure we have an automated test to catch this. Perhaps we should add one.

**Comment by jlippuner** [comment]
> > Looks great and great catch!
> > Since we're making changes to the boundary conditions file anyway, I thought this would be a good opportunity to reduce some code duplication. I've made one generic boundary condition function covering all 12 cases (X1, X2, X3) x (inner, outer) x (outflow, reflection). Hopefully, this would make it easier to avoid such copy&paste errors in the future (if the boundary condition code needs to change). Please let me know what you think.
> 
> In principal, I'm in favor of reduced code, but I worry that I can't quite parse whether or not the generic version covers all 12 specialized versions. I'm not sure we have an automated test to catch this. Perhaps we should add one.

We should definitely have a test. However, I think it should be relatively straightforward to check. There's only a few lines that were different between the expanded 12 cases. That being said, I'm happy to move the code duplication reduction to a new PR if it would take too long to review as part of this PR.

**Comment by Yurlungur** [comment]
>  That being said, I'm happy to move the code duplication reduction to a new PR if it would take too long to review as part of this PR.

That's up to @pgrete I suppose. My preference would be to add an automated test before merging that ensures these are working as intended. Advecting off of each face in 3d would be a. reasonable regression test, I think. We've clearly just been burned by our lack of testing in this area.

File: src/bvals/boundary_conditions.cpp
Comment (pgrete) [inline src/bvals/boundary_conditions.cpp]: For the sake of readability, how about using enum classes instead of the `bool`. Given that it's all compile time logic and bool/enum/int typically have the same size, it should also not affect performance.

File: src/bvals/boundary_conditions.cpp
Comment (pgrete) [inline src/bvals/boundary_conditions.cpp]:
```suggestion
  constexpr bool X1 = (DIR == X1DIR);
  constexpr bool X2 = (DIR == X2DIR);
  constexpr bool X3 = (DIR == X3DIR);
```

**Review by pgrete — COMMENTED** [review]
I like the reduced code (see also comment on readability).
I also think that extending the test coverage is a good idea.
I have an idea to use dc reconstruction with a cfl of 1, so that we an check inflow, outflow, reflect exactly as as there'd be no diffusion.

File: src/bvals/boundary_conditions.cpp
Comment (jlippuner) [reply src/bvals/boundary_conditions.cpp]: Good idea. I'll make that change

**Comment by jlippuner** [comment]
Waiting on test for boundary condition before merging

**Comment by kurtsansom** [comment]
What kind of test are you looking for? simple unit test exercising the templated boundary condition function?

**Comment by Yurlungur** [comment]
> What kind of test are you looking for? simple unit test exercising the templated boundary condition function?

Yes exactly. Just to make sure it behaves the same as the older non-templated code. (And/or is correct. ;) )

**Comment by pgrete** [comment]
This is now ready for new comments/review.
I now added checks for the new infrastructure (though user boundary conditions are not covered yet, see below).
I tried to be least intrusive to the advection example.
The key changes are:
- the FillDerived function now has a conditional that checks if all values are initialized (i.e., making sure the corners are filled, too)
- a new regression test, testing periodic, reflecting and outflow conditions enabled
  -  by migrating individual cells in all directions
  -  through an optional `v`(elocity) vector (so that reflecting boundaries are covered)
  - a new flux branch in the flux functions that translates cell data for cfl=1.0 (this works as the content of individual cells don't cross)file:///home/pgrete/test_reflect.mp4

  - the reflection test looks like: 

https://user-images.githubusercontent.com/22685568/133004699-1207cdd0-7073-4cb7-89ca-39be52fb38ef.mp4

Regarding the user boundary conditions:
In principle, I'd like to test all allowed permutations of boundary conditions as that combination of user and outflow threw me off in first place.
From a design point of view, we'd have to extend the advection example with 6 different user functions.
I imagine those just setting the `advected` field to a non-zero value and let the now implemented check in `FillDerived`/`SquareIt` handle it.
Do you think it's worth adding this change here or should we merge this sooner than later and take care of it in a separate PR?

**Comment by pgrete** [comment]
Ping @Yurlungur  for final re-review of latest changes as discussed on Thursday.

File: example/advection/advection_package.cpp
Comment (Yurlungur) [inline example/advection/advection_package.cpp]: What is this line supposed to be doing? Doesn't this return the const reference to the dictionary? Why the empty call?

File: example/advection/advection_package.cpp
Comment (Yurlungur) [inline example/advection/advection_package.cpp:325]: High value comment :+1:

File: example/advection/advection_package.cpp
Comment (Yurlungur) [inline example/advection/advection_package.cpp:503]: I like the `+=` syntax here. Feels nicer than the if-else block.

File: example/advection/advection_package.cpp
Comment (Yurlungur) [inline example/advection/advection_package.cpp]: rather than `+1` or `+2`, one could do `idx_v + XNDIR - 1` which might be a little cleaner. Not a big deal though.

File: example/advection/advection_package.cpp
Comment (Yurlungur) [inline example/advection/advection_package.cpp:61]: I don't love the name `block` for the test profile. But I don't have a better name either.

File: tst/regression/test_suites/bvals/bvals.py
Comment (Yurlungur) [inline tst/regression/test_suites/bvals/bvals.py]: I know it's default to reflecting, but I still might be explicit here, in case the parameter file changes.

**Review by Yurlungur — APPROVED** [review]
LGTM @pgrete some minor comments below. Really fantastic to have this test in place, along with the reduced code.

File: example/advection/advection_package.cpp
Comment (pgrete) [reply example/advection/advection_package.cpp]: Good catch. Looks like a debugging leftover. I'm removing it.

File: tst/regression/test_suites/bvals/bvals.py
Comment (pgrete) [reply tst/regression/test_suites/bvals/bvals.py]: Good point. Will do.

File: example/advection/advection_package.cpp
Comment (pgrete) [reply example/advection/advection_package.cpp]: Done.

**Comment by pgrete** [comment]
Addressed comment and triggering automerge now.

**Pull request merged**
